### PR TITLE
:sparkles: Version 0.5.0

### LINKED ISSUE #97: :sparkles: Add conformal prediction

### LINKED ISSUE #152: 🐛mc_dropout incompatible with mlp
The dropout layers in torch-uncertainty's [MLP model](https://github.com/ENSTA-U2IS-AI/torch-uncertainty/blob/main/torch_uncertainty/models/mlp.py) are implemented using `torch.functional`.

```python
    def forward(self, x: Tensor) -> Tensor | dict[str, Tensor]:
        for layer in self.layers:
            x = F.dropout(layer(x), p=self.dropout_rate, training=self.training)
            x = self.activation(x)
        return self.final_layer(x)
```

This means that a simple demo applying MC-dropout to an MLP:

```python
from torch_uncertainty.models.mlp import mlp
from torch_uncertainty.models import mc_dropout
....
model = mlp(
    in_features=2,
    num_outputs=datamodule.num_classes,
    hidden_dims=(3,3),
    dropout_rate=0.4,
)

mc_model = mc_dropout(model, num_estimators=16)
```
gives the error `ValueError: No dropout module found in the model. Please use nn.Dropout-like modules to apply dropout.`

### LINKED ISSUE #163: 🔨Transforms in Datamodules should be settable in the ``__init__()``
Allowing for overriding the default training and evaluation transforms in the datamodules would be more user-friendly.

### LINKED ISSUE #164: :hammer: [Post-processing] If `model` is None, use dl xs as logits
This would allow using the method without any inner model.

### LINKED ISSUE #165: :bug: Fix TemperatureScaler for binary problems

### LINKED ISSUE #167: 🔧 Enable direct imports
It would be nice to enable direct imports from top-level package without specifying submodules, especially for commonly used classes. For instance: 

```
from torch_uncertainty.metrics import CalibrationError
from torch_uncertainty.post_processing import TemperatureScaler
```

could become

`from torch_uncertainty import CalibrationError, TemperatureScaler`

Happy to contribute a PR if desired!

### LINKED ISSUE #176: :hammer: Put ECE in %
As suggested by @alafage.

### PR DESCRIPTION
**Pull request by o-laurent**
New Features
------------------

- Add Test-Time Adaptation and Zero (majority voting).
- Introduce `TUCheckpoint` callbacks, simplifying saving multiple checkpoints for Classification, Segmentation, and Regression routines.
- Scalers without models will work on the inputs provided by the dataloader.
- Add OOD detection support to `SegmentationRoutine`.
- Add Conformal post-hoc methods
- Add SGMCMC optimizers (i.e., SGLD and SGHMC)

QoL improvements
-------------------------

- lightning<2.5.0 is supported again!
- Add `eval_batch_size` parameter to all datamodules
- Enable overloading default train and test transforms in datamodules (resolve #163)
- `deep_ensembles` parameters can be stored on CPU and loaded one model at a time on the GPU during the forward pass
- The main bricks of code are available as second-level imports
- New documentation

Bug fixes & solved issues
------------

- Will fix #97 
- Will fix #152.
- Will fix #163
- Will fix #164
- Will fix #165
- Will fix #167
- Will fix #176

Todo
-------------------------

- [ ] Add all the new methods to the ReadMe
- [ ] Add the new tutorials to the ReadMe
- [x] Fix the coverage
- [ ] Collect all changes to prepare the release notes
- [ ] ~Remove device from post-processing~ will be done later

### CODE REVIEW

File: auto_tutorial_source/Post_Hoc_Methods/tutorial_conformal.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        v2.Normalize(mean=(0.4914, 0.4822, 0.4465), std=(0.2023, 0.1994, 0.2010)),
+    ]
+)
+
+
+datamodule = CIFAR10DataModule(
+    root="./data",
+    batch_size=batch_size,
+    num_workers=8,
+    eval_ood=True,
```
Comment (o-laurent) [inline auto_tutorial_source/Post_Hoc_Methods/tutorial_conformal.py:56]: Detail a bit

File: auto_tutorial_source/Post_Hoc_Methods/tutorial_conformal.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    val_split=0.2,
+)
+datamodule.setup()
+
+
+# %%
+# 3. Define training configuration (optimizer and scheduler)
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+def optim_recipe(model, lr_mult: float = 1.0):
```
Comment (o-laurent) [inline auto_tutorial_source/Post_Hoc_Methods/tutorial_conformal.py:65]: We don't need this

File: auto_tutorial_source/Post_Hoc_Methods/tutorial_conformal.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        for j in range(len(classes)):
+            ax.barh(classes[j], confidence_scores[i, j], color="blue")
+        ax.set_xlim(0, 1)
+        ax.set_xlabel("Confidence Score")
+    plt.tight_layout()
+    plt.show()
+
+
+# %%
+# 5. Calibrate with ConformalClsTHR
```
Comment (o-laurent) [inline auto_tutorial_source/Post_Hoc_Methods/tutorial_conformal.py:95]: I wouldn't put calibration which can be misunderstood

File: auto_tutorial_source/Post_Hoc_Methods/tutorial_conformal.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+confidence_scores = conformal_model.conformal(inputs.cuda())
+
+classes = datamodule.test.classes
+
+visualize_prediction_sets(inputs, labels, confidence_scores[:5].cpu(), classes)
+
+# %%
+# 7. Calibrate with ConformalClsAPS
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+print("\n[Phase 3]: ConformalClsAPS calibration")
```
Comment (o-laurent) [inline auto_tutorial_source/Post_Hoc_Methods/tutorial_conformal.py:126]: We don't need to skip lines

File: auto_tutorial_source/Regression/tutorial_probabilistic_regression.py [hunk 5]
```diff
@@ -83,24 +94,27 @@ def forward(self, x):
...
 from torch_uncertainty.losses import DistributionNLLLoss
 from torch_uncertainty.routines import RegressionRoutine
 
 loss = DistributionNLLLoss()
 
+
 def optim_regression(
     model: nn.Module,
-    learning_rate: float = 5e-3,
+    learning_rate: float = 5e-2,
```
Comment (o-laurent) [inline auto_tutorial_source/Regression/tutorial_probabilistic_regression.py:109]: We could remove the fct argument

File: auto_tutorial_source/Regression/tutorial_regression.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+model = mlp(in_features=6, num_outputs=1, hidden_dims=[10, 10])
+
+# %%
+# 4. The Loss and the Training Routine
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# This is a classification problem, and we use CrossEntropyLoss as the (negative-log-)likelihood.
+# We define the training routine using the classification routine from torch_uncertainty.routines.
+# We provide the number of classes, the model, the optimization recipe, the loss, and tell the routine
+# that our model is an ensemble at evaluation time with the `is_ensemble` flag.
```
Comment (o-laurent) [inline auto_tutorial_source/Regression/tutorial_regression.py:67]: To get the corresponding metrics

File: torch_uncertainty/models/segmentation/segformer.py [hunk 7]
```diff
@@ -387,46 +387,46 @@ def _init_weights(self, m):
...
+            inputs = blk(inputs, h, w)
+        inputs = self.norm4(inputs)
+        inputs = inputs.reshape(b, h, w, -1).permute(0, 3, 1, 2).contiguous()
+        outs.append(inputs)
 
         return outs
 
-    def forward(self, x):
-        return self.forward_features(x)
+    def forward(self, inputs):
```
Comment (o-laurent) [inline torch_uncertainty/models/segmentation/segformer.py:428]: missing types

File: auto_tutorial_source/Classification/tutorial_distribution_shift.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+# %%
+# DataModule Setup
+# ----------------
+#
+# TorchUncertainty provides convenient DataModules for standard datasets like CIFAR-10.
+# DataModules handle data loading, preprocessing, and batching, simplifying the data pipeline. Each datamodule
+# also include the corresponding out-of-distribution and distribution shift datasets, which are then used by the routine.
+# For CIFAR-10, the corresponding distribution-shift dataset is CIFAR-10C as used in the community.
+# To enable OOD evaluation, activate the `eval_shift` flag as done below.
```
[the remark this answers is not in the document]
Comment (alafage) [reply auto_tutorial_source/Classification/tutorial_distribution_shift.py:38]: Yes, but in this case we want to do Distribution Shift evaluation

File: auto_tutorial_source/Classification/tutorial_bayesian.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+trainer.fit(model=routine, datamodule=datamodule)
+trainer.test(model=routine, datamodule=datamodule)
+# %%
+# 6. Testing the Model
+# ~~~~~~~~~~~~~~~~~~~~
+#
+# Now that the model is trained, let's test it on MNIST.
+# Please note that we apply a reshape to the logits to determine the dimension corresponding to the ensemble
+# and to the batch. As for TorchUncertainty 0.4.3, the ensemble dimension is merged with the batch dimension
```
[the remark this answers is not in the document]
Comment (toto-hash) [reply auto_tutorial_source/Classification/tutorial_bayesian.py:135]: Maybe say that it's like that for the output of the routine? The output shape of ensemble models is different.

[the remark this answers is not in the document]
Comment (o-laurent) [reply auto_tutorial_source/Classification/tutorial_bayesian.py:135]: State that it's true for the output of the routine maybe? The shape of the model output is different so it may be misleading?

File: auto_tutorial_source/Classification/tutorial_distribution_shift.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (o-laurent) [reply auto_tutorial_source/Classification/tutorial_distribution_shift.py:38]: True, but the sentence makes no sense anyway

File: auto_tutorial_source/Regression/tutorial_probabilistic_regression.py (hunk 5 above)
Comment (o-laurent) [reply auto_tutorial_source/Regression/tutorial_probabilistic_regression.py:109]: 💯

File: docs/source/api.rst [hunk 10]
```diff
@@ -416,9 +442,10 @@ Classification
 
     CIFAR10DataModule
     CIFAR100DataModule
+    ImageNetDataModule
     MNISTDataModule
     TinyImageNetDataModule
-    ImageNetDataModule
+    
 
 UCI Tabular Classification
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply docs/source/api.rst:450]: Let's postpone that, you re right

File: pyproject.toml [hunk 11]
```diff
@@ -128,39 +163,16 @@ lint.ignore = [
...
-    "build",
-    "dist",
-    "venv",
-]
-
-[tool.ruff.lint.pydocstyle]
-convention = "google"
+[tool.ruff.format]
+docstring-code-format = true
+docstring-code-line-length = 80
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply pyproject.toml:175]: After discussions with @alafage postponed to last PR commit

File: torch_uncertainty/datasets/utils.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        dataset (Dataset): The dataset to be split.
+        val_split_rate (float): The amount of the original dataset to use as validation split.
+        val_transforms (Callable | None, optional): The transformations to apply on the validation set.
+            Defaults to ``None``.
+
+    Returns:
+        tuple[Dataset, Dataset]: The training and the validation splits.
+    """
+    train, val = random_split(dataset, [1 - val_split_rate, val_split_rate])
+    val = copy.deepcopy(val)
```
[the remark this answers is not in the document]
Comment (alafage) [reply torch_uncertainty/datasets/utils.py:25]: It is to prevent modifying the original transform of the dataset. And thus modifying the train transform.

File: torch_uncertainty/models/wrappers/swag.py [hunk 13]
```diff
@@ -233,10 +232,10 @@ def load_state_dict(self, state_dict: Mapping, strict: bool = True, assign: bool
         self._load_swag_stats(state_dict)
         return super().load_state_dict(state_dict, strict, assign)
 
-    def compute_logdet(self, block=False):
+    def compute_logdet(self, block: bool = False) -> NoReturn:
```
[the remark this answers is not in the document]
Comment (alafage) [reply torch_uncertainty/models/wrappers/swag.py:235]: yes. putting ``None`` instead

File: torch_uncertainty/routines/segmentation.py [hunk 14]
```diff
@@ -311,6 +412,18 @@ def log_segmentation_plots(self) -> None:
                 show(pred_mask, gt_mask),
             )
 
+    def save_results_to_csv(self, results: dict[str, float]) -> None:
```
[the remark this answers is not in the document]
Comment (alafage) [reply torch_uncertainty/routines/segmentation.py:415]: Why though? We could delete it and do the saving in on_test_epoch_end

[the remark this answers is not in the document]
Comment (o-laurent) [reply torch_uncertainty/routines/segmentation.py:419]: It's just that it's quite generic and is copied between segmentation & classification

File: torch_uncertainty/post_processing/conformal/conformal_aps.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+        logits = self.model(inputs.to(self.device))
+        return self.transform(logits)
+
+    def _sort_sum(self, probs: Tensor):
+        """Sort probabilities and compute cumulative sums."""
+        ordered, indices = torch.sort(probs, dim=-1, descending=True)
+        cumsum = torch.cumsum(ordered, dim=-1)
+        return indices, ordered, cumsum
+
+    def _calculate_single_label(self, probs: Tensor, labels: Tensor):
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply torch_uncertainty/post_processing/conformal/conformal_aps.py:62]: I don't know what it means so let's keep it this way

File: torch_uncertainty/metrics/classification/mean_iou.py [hunk 16]
```diff
@@ -35,14 +35,14 @@ def __init__(
         Shape:
             As input to ``forward`` and ``update`` the metric accepts the following input:
 
-            - ``preds`` (:class:`~torch.Tensor`): An int tensor of shape ``(B, ...)`` or float tensor of shape ``(B, C, ..)``.
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply torch_uncertainty/metrics/classification/mean_iou.py:38]: @alafage the * instead of `, is it the way it's supposed to be?

File: torch_uncertainty/utils/cli.py [hunk 17]
```diff
@@ -125,5 +156,9 @@ def add_default_arguments_to_parser(self, parser: LightningArgumentParser) -> No
 
     def add_arguments_to_parser(self, parser: LightningArgumentParser) -> None:
         super().add_arguments_to_parser(parser)
-        parser.link_arguments("data.eval_ood", "model.eval_ood")
-        parser.link_arguments("data.eval_shift", "model.eval_shift")
+        with contextlib.suppress(ValueError):
+            parser.link_arguments("data.eval_ood", "model.eval_ood")
```
[the remark this answers is not in the document]
Comment (o-laurent) [reply torch_uncertainty/utils/cli.py:160]: Let's keep it for later

File: pyproject.toml (hunk 11 above)
[the remark this answers is not in the document]
Comment (o-laurent) [reply pyproject.toml:167]: My bad, I read too fast.

File: auto_tutorial_source/Bayesian_Methods/tutorial_mc_batch_norm.py [hunk 18]
```diff
@@ -85,14 +87,12 @@
 # predictions. We specify that the BatchNorm layers are to be converted to
 # MCBatchNorm layers, and that we want to use 8 stochastic estimators.
 # The amount of stochasticity is controlled by the ``mc_batch_size`` argument.
-# The larger the ``mc_batch_size``, the more stochastic the predictions will be.
-# The authors suggest 32 as a good value for ``mc_batch_size`` but we use 16 here
+# The larger the ``mc_batch_size``, the less stochastic the predictions will be.
+# The authors suggest 32 as a good value for ``mc_batch_size`` but we use 4 here
```
Comment (alafage) [inline auto_tutorial_source/Bayesian_Methods/tutorial_mc_batch_norm.py:91]: `mc_batch_size=16` not `4`

**Review by alafage — APPROVED** [review]
